src/lease.ts

```diff
@@ -0,0 +1,255 @@
+        this.emit('lost', err);
+        // return, don't throw, from here so that if no one is listening to
+        // grant() we don't crash the process.
+        return err;
+      });
+  }
+
+  /**
+   * Grant waits for the lease to be granted. You generally don't need to
+   * call this, as any operations with `.put` will queue automatically.
+   *
+   * Calling this multiple times is safe; it won't try to request multipl leases.
+   *
+   * It rejects if the lease cannot be granted, in additon to the `lost`
+   * event firing.
+   */
+  public grant(): Promise<string> {
+    return this.leaseID.then(throwIfError);
+  }
+
+  /**
+   * Revoke frees the lease from etcd. Keys that the lease owns will be
+   * evicted.
+   */
+  public revoke(): Promise<void> {
+    this.close();
+    return this.leaseID.then(id => {
+      if (!(id instanceof Error)) { // if an error, we didn't grant in the first place
+        return this.client.leaseRevoke({ ID: id }).then(() => undefined);
+      }
+
+      return undefined;
+    });
+  }
+
+  /**
+   * Put returns a put builder that operates within the current lease.
+   */
+  public put(key: string | Buffer): PutBuilder {
+    return new PutBuilder(new RPC.KVClient(new LeaseClientWrapper(this.leaseID, this.pool)), key);
+  }
+
+  /**
+   * keepaliveOnce fires an immediate keepalive for the lease.
+   */
+  public keepaliveOnce(): Promise<RPC.ILeaseKeepAliveResponse> {
+    return Promise.all([
+      this.client.leaseKeepAlive(),
+      this.grant(),
+    ]).then(([stream, id]) => {
+      return new Promise<RPC.ILeaseKeepAliveResponse>((resolve, reject) => {
+        stream.on('data', res => resolve(res));
+        stream.on('error', err => reject(castGrpcError(err)));
+        stream.write({ ID: id });
+      })
+      .then(res => {
+        stream.end();
+        if (leaseExpired(res)) {
+          const err = new EtcdLeaseInvalidError(res.ID);
+          this.emit('lost', err);
+          this.close();
+          throw err;
+        }
+
+        return res;
+      });
+    });
+  }
+
+  /**
+   * Returns whether etcd has told us that this lease revoked.
+   */
+  public revoked(): boolean {
+    return this.state === State.Revoked;
+  }
+
+  /**
+   * A `lost` event is fired when etcd indicates that we've lost the lease
+   * on this client. This can be a result of a number of events:
+   *  - We've not been able to contact etcd for a while and our TTL has
+   *    definitely expired (emits a EtcdLeaseInvalidError)
+   *  - We contacted etcd and it said that the lease was expired, or revoked
+   *    (emits a EtcdLeaseInvalidError).
+   *  - We weren't able to get an initial grant for the lease.
+   * This is NOT fired if `revoke()` is called manually.
+   */
+  public on(event: 'lost', handler: (err: EtcdError) => void): this;
+
+  /**
+   * keepaliveFired is emitted whenever we start
+   * trying to send a lease keepalive.
+   */
+  public on(event: 'keepaliveFired', handler: () => void): this;
```

Comment:
I used a pattern in one of my personal projects which may (or may not) help.
https://github.com/SimonSchick/EliteDangerousUtils/blob/master/src/EDLog/EDLog.ts#L261

I hope the TS team comes up with a better solution with methods taking the same arguments etc.